content/blog/2019-11-13-efficient-path-prof/index.md

```diff
@@ -0,0 +1,205 @@
+
+<img src="simple_inc.png" style="width: 30%">
+
+All `Inc` mappings are required to have the same sum on a path `P` through the CFG as the sum of the `Val` mappings for path `P`. The edge with `Inc(e) = -2)` for the edge from A to B was computed by taking the path from B to A in the spanning tree. `Val((A,B))` was 2 but because we traversed the edge backwards, we subtracted 2 from `Inc` of the other edge.
+
+Finally, the algorithm inserts instrumentation according to the `Inc` mapping. For every edge in the original CFG with a non-zero `Inc` value, the instruction `r += Inc(e)` is inserted by creating a new basic block between the source and destination of the edge with that instruction. The block is then inserted into the CFG by rerouting the source's successor to the new block and adding a jump at the end of the new block to the destination block. The entry block adds an instruction to initialize `r=0` and the exit block adds an instruction `path_table[r++]`. Back edges are bit more interesting. For back edges we insert the instruction sequence:
+
+```
+r += backedge_inc
+path_table[r++]
+r = backedge_reset
+```
+
+`backedge_inc` is set to `Inc((entry,backedge.dest))`, which is the `Inc` value of the edge from the entry block to the head of the loop containing the back edge. `backedge_reset` is set to `Inc((backedge.source, exit))`, which is the `Inc` value of the edge from the bottom of the loop to the exit block. This was somewhat underspecified in the paper but I believe it makes sense based on their descriptions of what the two inserted edges for each back edge represent. This is the instrumented CFG of the example we have been looking at:
+
+<img src="simple_inst.png" style="width: 50%">
+
+We can verify that our path mappings are preserved by figuring out the value of `r` at the end of every path before we do `path_table[r]++`:
+
+| Path | r  |
+|:-:|:-:|
+| A,B,D  | 0 |
+| A,B,C  | 1 |
+| B,C  | 3 |
+| B,D  | 2 |
+
+Also, note that this matches the `Val` sum for each path computed above.
+
+### Differences between the paper
```

Comment:
Differences *from* the paper? Or differences between the paper *and this implementation*?